### PR TITLE
p2p/net/conn: reuse listen tcp port to dial

### PR DESCRIPTION
**Pull request by jbenet**
This PR finally makes it possible to dial out from our TCP listen addr. I had to sink low into net to make this possible. The result is the go-reuseport package. Sadly `golang/net` doesnt let us add tcp socket options, so we have to do it this way (setting up the sockets manually, which is truly painful). can see the same approach used in various modules:
- https://github.com/kavu/go_reuseport
- https://github.com/benburkert/net.tune/ [1]

Anyway, this PR:
- adds go-reuseport (and its dep, go-sockaddr/net)
- changes p2p/net/conn to optimistically use go-reuseport (it can't always work because we may be dialing the same 5-tuple), or fallback to reg dial.
- changed the identify receiver to register the observed address, if and only if it is for a connection whose LocalAddress() is one of our ListenAddrs().

[1] wow i just discovered this one... after rolling my own. fffff! it looks good-- we might want to switch to this one if my go-reuseport gives us problems

### CODE REVIEW

**Comment by jbenet** [comment]
https://build.protocol-dev.com/job/ipfs_test_all_commits_on_branch/185/console <-- tells me tests_all_commits fails, but i cant tell which commit. `make test_all_commits` passes locally

**Comment by jbenet** [comment]
RFCR

**Comment by whyrusleeping** [comment]
this PR shouldnt be against ipfs, you should file it against Go.

File: Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+	if err == nil {
+		syscall.CloseOnExec(fd)
+	}
+	syscall.ForkLock.RUnlock()
+
+	if err != nil {
+		return -1, err
+	}
+
+	// set non-blocking until after connect, because we cant poll using runtime :(
```
Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: whats this?

File: Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+		// but of course, it uses fucking runtime_* functions that _cannot_ be used by
+		// non-go-stdlib source... seriously guys, what kind of bullshit is that!?
+		<-time.After(20 * time.Microsecond)
+		var nerr int
+		nerr, err = syscall.GetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_ERROR)
+		if err != nil {
+			return err
+		}
+		switch err = syscall.Errno(nerr); err {
+		case syscall.EINPROGRESS, syscall.EALREADY, syscall.EINTR:
```
Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: possibly `EAGAIN` as well?

File: Godeps/_workspace/src/github.com/jbenet/go-sockaddr/net/net.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+import (
+	"net"
+	"syscall"
+)
+
+// NetAddrAF returns the syscall AF_* type for a given net.Addr
+// returns AF_UNSPEC if unknown
+func NetAddrAF(addr net.Addr) int {
+	switch addr := addr.(type) {
+	default:
```
Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/jbenet/go-sockaddr/net/net.go]: seeing default at the top of a switch always weirds me out

File: p2p/net/conn/dial.go [hunk 4]
```diff
@@ -92,6 +70,78 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
...
+	if laddr != nil {
+		// dial using reuseport.Dialer, because we're probably reusing addrs.
+		// this is optimistic, as the reuseDial may fail to bind the port.
+		log.Debugf("trying to reuse: %s", laddr)
+		if nconn, err := d.reuseDial(laddr, raddr); err == nil {
+			// if it worked, wrap the raw net.Conn with our manet.Conn
+			log.Debugf("reuse worked! %s %s %s", laddr, nconn.RemoteAddr(), nconn)
+			return manet.WrapNetConn(nconn)
+		}
+		// if not, we fall back to regular Dial without a local addr specified.
```
Comment (whyrusleeping) [inline p2p/net/conn/dial.go]: Can we log when this fails? Might be good info to see

**Comment by whyrusleeping** [comment]
Is much of this code directly copied from go's `net` lib? If so, it might be good to mark any of that as such.

**Comment by jbenet** [comment]
> this PR shouldnt be against ipfs, you should file it against Go.

Right!?

> Is much of this code directly copied from go's `net` lib? If so, it might be good to mark any of that as such.

Yeah, the stuff copied straight out is marked with the copyright notice. The `connect()` function is a stripped down version of theirs.

File: p2p/net/conn/dial.go (hunk 4 above)
Comment (jbenet) [inline p2p/net/conn/dial.go]: I originally had a log here, but seemed like it was a lot of noise. I'll remove the "trying to reuse" one.

And, i think it's happening more often than I expect. I think I'm not handling something right and so the `connect()` call fails when it shouldn't have to

File: Godeps/_workspace/src/github.com/jbenet/go-sockaddr/net/net.go (hunk 3 above)
Comment (jbenet) [inline Godeps/_workspace/src/github.com/jbenet/go-sockaddr/net/net.go]: fixed in 66d375b

**Comment by jbenet** [comment]
This amount of copy pasting is horrible. This PR is an exercise in why internal packages / forcibly unaccessible symbols are a bad idea. There's a significant difference between "not exporting something to my interface" and "making it absolutely impossible for you to use this code". It should be possible to import things _if you really mean it_. Maybe as a different package altogether with a dependency like ibaz "foo/bar/baz/internal".

**Comment by jbenet** [comment]
https://build.protocol-dev.com/job/build%20docker%20image/923/console

```
Sending build context to Docker daemon 
Step 0 : FROM golang:1.3
 ---> ebd3fd90ae2e
Step 1 : MAINTAINER Brian Tiger Chow <btc@perfmode.com>
 ---> Using cache
 ---> 6c893fd2f62d
Step 2 : ADD . /go/src/github.com/jbenet/go-ipfs
 ---> a18697f6bce6
Removing intermediate container 764cfa2d8ed1
Step 3 : RUN cd /go/src/github.com/jbenet/go-ipfs/cmd/ipfs && go install
 ---> Running in 3452c2eb3913
 ---> 89b1045e2b93
Removing intermediate container 3452c2eb3913
Step 4 : EXPOSE 4001 5001 4002/udp
 ---> Running in ee9c69101f50
 ---> 2c6dc0ce7374
Removing intermediate container ee9c69101f50
Step 5 : ENTRYPOINT ipfs
 ---> Running in b5429cafd23b
2015/01/20 00:14:49 An error occured while creating the container
Build step 'Execute shell' marked build as failure
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Triggering a new build of chore_reclaim-docker-space
Finished: FAILURE
```
Maybe the chore should happen before the docker image tries to build and fails? or this may be unrelated.

cc @briantigerchow

File: Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go (hunk 2 above)
Comment (jbenet) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: Hm, what makes you think so? It isn't in: http://golang.org/src/net/fd_unix.go?h=connect#L71 -- which is where i got this from.

Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: `EAGAIN` is returned when there is no data available during non-blocking socket IO, it just seemed like something that might come up here. But if youre pulling these from official Go source, its probably best to leave this as is.

Comment (jbenet) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: yeah. good instinct. in this case looks like `EAGAIN` on `connect()` in linux is a larger connect problem: http://man7.org/linux/man-pages/man2/connect.2.html#EAGAIN -- not the usual R/W `EAGAIN`

Comment (whyrusleeping) [inline Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go]: interesting.... yeah.. sockets are shitty. the future is ipfs datastreams:

```
stream := ipfs.StreamTo(peerid)
stream.Write("some data")
```

**Comment by whyrusleeping** [comment]
Looks pretty good to me, let's hope it solves our issues!

**Comment by jbenet** [comment]
@whyrusleeping is that a LGTM on the PR?

**Comment by whyrusleeping** [comment]
Yeah, merge it! (before i change my mind! :P )

**Pull request merged**